storage/transfermanager/downloader.go

```diff
@@ -402,6 +525,43 @@ func (in *DownloadObjectInput) downloadShard(client *storage.Client, timeout tim
 	return
 }
 
+// DownloadDirectoryInput is the input for a directory to download.
+type DownloadDirectoryInput struct {
+	// Bucket is the bucket in GCS to download from. Required.
+	Bucket string
+	// LocalDirectory specifies the directory to download the matched objects
+	// to. Relative paths are allowed. The directory will be created if it does
+	// not already exist. Required.
+	LocalDirectory string
+
+	// Prefix is the prefix filter to download objects whose names begin with this.
+	// Optional.
+	Prefix string
+
+	// StartOffset is used to filter results to objects whose names are
+	// lexicographically equal to or after startOffset. If endOffset is also set,
+	// the objects listed will have names between startOffset (inclusive) and
+	// endOffset (exclusive).
+	StartOffset string
+
+	// EndOffset is used to filter results to objects whose names are
+	// lexicographically before endOffset. If startOffset is also set, the
+	// objects listed will have names between startOffset (inclusive) and
+	// endOffset (exclusive). Optional.
+	EndOffset string
+
+	// MatchGlob is a glob pattern used to filter results (for example, foo*bar). See
+	// https://cloud.google.com/storage/docs/json_api/v1/objects/list#list-object-glob
+	// for syntax details. Optional.
+	MatchGlob string
+
+	// Callback will run after each object in the directory as selected by the
+	// provided filters is finished downloading. It must be set if and only if
+	// the [WithCallbacks] option is set.
+	// WaitAndClose will wait for all callbacks to finish.
+	Callback func(*DownloadOutput)
```

Comment:
One more thought about callbacks. Thinking about it some more, it seems like if someone is running in async mode (not calling WantAndClose) they would not have a way to know when the whole directory download is complete. Any thoughts about how to address this? Perhaps it makes more sense to have the callback take []DownloadOutput and only trigger once the whole directory is finished?

Reply:
Hmmm... good point. I think it makes sense for the callback to be called on each file download. But I agree there should be a way of knowing when the directory is completely downloaded. 

What do you think of changing the current callbacks into a `ProgressFunc` (since keeping the code to trigger it would be pretty simple and it's a useful feature) and then have `Callback` be as you describe? I'll open another PR for this since it will require a bit more involved code added.